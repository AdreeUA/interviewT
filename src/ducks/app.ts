import { createAction, createReducer } from 'redux-act'

export const REDUCER = 'APP'
const NS = `${REDUCER}__`

export interface AppState {
  isLoading: boolean
}

export const initialState: AppState = {
  isLoading: false,
}

const reducer = createReducer({}, initialState)

export const setLoading = createAction<boolean>(`${NS}SET`)
reducer.on(setLoading, (state, isLoading) => ({ ...state, isLoading }))

export default reducer
