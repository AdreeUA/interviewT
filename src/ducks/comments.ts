import { createAction, createReducer } from 'redux-act'
import axios from 'axios'
import { ReduxThunkCb } from '../store/types'

export const REDUCER = 'COMMENTS'
const NS = `${REDUCER}__`

export interface CommentsState {
  items: {
    body: string
    email: string
    id: number
    name: string
    postId: number
  }[]
}

export const initialState: CommentsState = {
  items: []
}

const reducer = createReducer({}, initialState)

const set = createAction<any>(`${NS}SET`)
reducer.on(set, (state, items) => ({ ...state, items: [...items],  }))

export const load = (postId: number): ReduxThunkCb => dispatch => {
  return axios.get(`/comments?postId=${postId}`).then(response => {
    dispatch(set(response.data))
  })
}

export default reducer
