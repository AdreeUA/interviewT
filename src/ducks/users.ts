import { createAction, createReducer } from 'redux-act'
import axios from 'axios'
import { ReduxThunkCb } from '../store/types'

export const REDUCER = 'USERS'
const NS = `${REDUCER}__`

export interface UserState {
  thumbnail: string
  fullName: string
  email: string
  phone: string
}

export interface UsersState {
  results: UserState[]
  filtersResults: UserState[]
  filtersData: { [key: string]: string }
  isRequestError: boolean
  isRequesting: boolean
}

export const initialState: UsersState = {
  results: [],
  filtersResults: [],
  filtersData: {},
  isRequestError: false,
  isRequesting: false,
}

const reducer = createReducer({}, initialState)

const callFailure = createAction(`${NS}CALL_FAILURE`)
reducer.on(callFailure, (state) => ({
  ...state,
  isRequestError: true,
  isRequesting: false,
}))

const callSuccess = createAction(`${NS}CALL_SUCCESS`)
reducer.on(callSuccess, (state) => ({
  ...state,
  isRequestError: false,
  isRequesting: false,
}))

const callRequest = createAction(`${NS}CALL_REQUEST`)
reducer.on(callRequest, (state) => ({
  ...state,
  isRequestError: false,
  isRequesting: true,
}))

const set = createAction<any>(`${NS}SET`)
reducer.on(set, (state, users) => ({ ...state, results: users.results }))

const add = createAction<any>(`${NS}ADD`)
reducer.on(add, (state, { results }) => {
  const user = results[0]
  const nameField = user.name
  const titleName = nameField.title
  const firstName = nameField.first
  const lastName = nameField.last
  const fullName = `${titleName} ${firstName} ${lastName}`
  const thumbnail = user.picture.thumbnail
  const email = user.email
  const phone = user.phone
  return {
    ...state,
    results: [ ...state.results, { thumbnail, fullName, email, phone } ],
  }
})

export const setFilterData = createAction<any>(`${NS}SET_FILTER_DATA`)
reducer.on(setFilterData, (state, filterData) => {
  return {
    ...state,
    filtersData: {
      ...state.filtersData,
      ...filterData,
    },
  }
})

export const filterUsers = createAction(`${NS}FILTERS`)
reducer.on(filterUsers, (state) => {
  const filterKey = Object.keys(state.filtersData)
  const filtersResults = state.results.filter((user) => {
    const filterCollection: boolean[] = []
    filterKey.forEach((key) => {
      const result = user[key].toLowerCase().includes(state.filtersData[key].toLowerCase())
      return filterCollection.push(result)
    })
    return filterCollection.every((el: boolean) => el)
  })
  return {
    ...state,
    filtersResults,
  }
})

export const load = (): ReduxThunkCb => (dispatch) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest())
    return axios
      .get('https://randomuser.me/api/')
      .then((response) => {
        dispatch(add(response.data))
        dispatch(filterUsers())
        dispatch(callSuccess())
        resolve()
      })
      .catch((error) => {
        dispatch(callFailure())
        reject(error)
      })
  })
}

export default reducer
