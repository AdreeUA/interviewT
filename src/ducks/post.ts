import { createAction, createReducer } from 'redux-act'
import { matchPath } from 'react-router'
import axios from 'axios'
import { ReduxThunkCb } from '../store/types'
import { load as loadComments } from './comments'
import * as _ from 'lodash'

export const REDUCER = 'POST'
const NS = `${REDUCER}__`

export interface PostState {
    body: string
    id: number
    title: string
    userId: number
}

export const initialState: PostState | {} = {
}

const reducer = createReducer({}, initialState)

const set = createAction<{}>(`${NS}SET`)
reducer.on(set, (state, post) => ({ ...state, ...post }))

export const load = (): ReduxThunkCb => (dispatch, getState) => {
  const state = getState()
  const match = matchPath(state.router.location.pathname, { path: '/posts/:id' })
  const id = _.get(match, 'params.id', null)
  if (_.isNull(id)) {
    return Promise.reject(new Error('id must have a value'))
  }
  return axios.get(`/posts/${id}`).then(response => {
    dispatch(set(response.data))
    return dispatch(loadComments(id))
  })
}

export default reducer
