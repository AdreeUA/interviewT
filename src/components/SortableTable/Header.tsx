import React from 'react'
import { Dispatch, connect } from 'react-redux'
import { StoreState } from '../../store/types'
import { setFilterData, filterUsers } from '../../ducks/users'

interface OwnProps {}

interface ConnectedState {}

interface ConnectedDispatch {
  dispatch: Dispatch<StoreState>
}

type Props = OwnProps & ConnectedState & ConnectedDispatch

interface State {}

const mapStateToProps = (state: StoreState) => ({})

const WrappedHeader = connect<ConnectedState, ConnectedDispatch, OwnProps>(mapStateToProps)(
  class Header extends React.Component<Props, State> {
    tableColumns: { label: string; key: string; sort?: boolean }[] = [
      {
        label: 'Image',
        key: 'thumbnail',
      },
      {
        label: 'Name',
        sort: true,
        key: 'fullName',
      },
      {
        label: 'Email',
        sort: true,
        key: 'email',
      },
      {
        label: 'Phone',
        sort: true,
        key: 'phone',
      },
    ]

    handleChangeSearch = (key: string) => (event: React.SyntheticEvent<any>) => {
      const value = event.currentTarget.value
      this.props.dispatch(setFilterData({ [key]: value }))
      this.props.dispatch(filterUsers())
    }

    render() {
      return (
        <thead>
          <tr>
            {this.tableColumns.map((item) => {
              return (
                <th key={item.key} className={`table__header-${item.key}`}>
                  {item.label}
                  {item.sort && <input className="table__input" onChange={this.handleChangeSearch(item.key)} />}
                </th>
              )
            })}
          </tr>
        </thead>
      )
    }
  },
)

export default WrappedHeader
