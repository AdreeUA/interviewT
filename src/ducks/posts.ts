import { createAction, createReducer } from 'redux-act'
import axios from 'axios'
import { ReduxThunkCb } from '../store/types'
import { PostState } from './post'

export const REDUCER = 'POSTS'
const NS = `${REDUCER}__`

export interface PostsState {
  items: PostState[]
}

export const initialState: PostsState = {
  items: [],
}

const reducer = createReducer({}, initialState)

const set = createAction<any[]>(`${NS}SET`)
reducer.on(set, (state, items) => ({ ...state, items: [...items] }))

export const load = (): ReduxThunkCb => dispatch => {
  return axios.get('/posts').then(response => {
    dispatch(set(response.data))
  })
}

export default reducer
