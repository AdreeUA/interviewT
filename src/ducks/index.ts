import { combineReducers } from 'redux'
import app from './app'
import posts from './posts'
import post from './post'
import comments from './comments'
import users from './users'

export default combineReducers({
  app,
  posts,
  post,
  comments,
  users
})
