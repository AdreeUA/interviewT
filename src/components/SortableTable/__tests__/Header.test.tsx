import * as React from 'react'
import thunk from 'redux-thunk'
import { shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { setFilterData, filterUsers } from '../../../ducks/users'
import Header from '../Header'

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

jest.mock('../../../ducks/users', () => {
  return {
    setFilterData: jest.fn(() => {
      return () => Promise.resolve()
    }),
    filterUsers: jest.fn(() => {
      return () => Promise.resolve()
    }),
  }
})

describe('Header', () => {
  it('render() with initialState', () => {
    const store = mockStore({})
    const wrapper = shallow(<Header />, { context: { store } })
    expect(wrapper.dive()).toMatchSnapshot()
  })
  it('Input onChange()', () => {
    const store = mockStore({})
    const wrapper = shallow(<Header />, { context: { store } })
    const input = wrapper.dive().find('input').first()
    input.simulate('change', { currentTarget: { value: 'Site' } })
    expect(setFilterData).toBeCalledWith({ fullName: 'Site' })
    expect(filterUsers).toBeCalled()
  })
})
