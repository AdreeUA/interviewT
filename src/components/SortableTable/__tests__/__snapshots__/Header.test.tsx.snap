// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render() with initialState 1`] = `
<thead>
  <tr>
    <th
      className="table__header-thumbnail"
      key="thumbnail"
    >
      Image
    </th>
    <th
      className="table__header-fullName"
      key="fullName"
    >
      Name
      <input
        className="table__input"
        onChange={[Function]}
      />
    </th>
    <th
      className="table__header-email"
      key="email"
    >
      Email
      <input
        className="table__input"
        onChange={[Function]}
      />
    </th>
    <th
      className="table__header-phone"
      key="phone"
    >
      Phone
      <input
        className="table__input"
        onChange={[Function]}
      />
    </th>
  </tr>
</thead>
`;
